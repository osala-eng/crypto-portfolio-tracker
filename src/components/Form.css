.form-container {
    display: flex;
    justify-content: center;
    align-items: center;
    border: 2px solid #000;
    border-radius: 20px;
    margin: 50px auto;
}

.input-line {
    margin: 20px;
}

#username_field {
    width: 200px;
    height: 20px;
}

#input-form {
    display: flex;
    flex-direction: column;
}

#submit_button {
    width: 100px;
    height: 40px;
}