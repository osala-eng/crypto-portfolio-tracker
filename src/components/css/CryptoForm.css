#crypto-input-form {
    display: flex;
    flex-direction: row;
    max-width: 600px;
    align-items: center;
    justify-content: space-between;
    margin-top: 20px;
    position: relative;
}

.crypto-input-field{
    display: flex;
    flex-direction: row;
    align-items: center;
}
#crypto-button-container{
    display: flex;
    flex-direction: column;
    align-items: center;
    position: relative;
    top: 0;
}
.crypto-text-input-class{
    height: 20px;
    padding-left: 5px;
    padding-right: 5px;
    font-size: 14px;
    max-width: 130px;
    margin-left: 20px;
}
#dashboard_add_button{
    width: 100px;
    height: 30px;
    cursor: pointer;
}

#add_asset_error{
    color: red;
    position: absolute;
    bottom: -25px;
    min-width: 300px;
}
