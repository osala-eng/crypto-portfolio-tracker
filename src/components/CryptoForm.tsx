import React, { useEffect, useState } from 'react';
import { ChangeEvent, ClickEvent, HTTP, ID, MS } from '../data/types';
import { AssetsService } from '../data/config';
import Emitter from './Emitter';
import './css/CryptoForm.css';


export const CryptoForm = ({
    username, update, test=false }: {
    username?: string,
    test?: boolean
    update?: ()=>void }) => {
    const [error, setError] = useState(false);
    const [loading, setLoading] = useState(false);
    const [state, setState] = useState({
        username: username,
        token: '',
        quantity: NaN });
    const [edit, setEdit] = useState(false)

    /* istanbul ignore next */
    const handleSubmit = async (event: ClickEvent) => {
        event.preventDefault();
        if (state.quantity < 0 || !state.username || !state.token.length) {
            setError(true);
        }
        else {
            setLoading(true);
            await fetch(AssetsService, {
                method: 'POST',
                body: JSON.stringify(state)})
                .then(res => {
                    setLoading(false);
                    if (res.status !== HTTP['200']) {
                        throw new Error('Operation failed');
                    }
                    setState({...state, token: '', quantity: NaN});
                    // Update tokens table on success
                    update!();
                    setEdit(false);
                })
                .catch(() => {
                    setError(true);
                    setState({...state, token: '', quantity: NaN});
                    setEdit(false);
                });
        }
    };


    /* istanbul ignore next */
    useEffect(() => {
        const timer = setTimeout(() => {
            setError(false);
        }, MS['10000']);

        const listner = Emitter.addListener('edit-assets',
            ({token, quantity}:{token: string, quantity: number})=>{
                setEdit(true);
                setState( s => { return {...s,  token, quantity}})
        });

        return () => {
            clearTimeout(timer);
            listner.remove();
        }
    }, [error]);

    /* istanbul ignore next */
    const changeEvent = (event: ChangeEvent, id: number) => {
        if (id === ID['0']) {
            setState({ ...state, token: event.target.value });
        }
        else if (id === ID['1']) {
            setState({ ...state, quantity: +event.target.value});
        }
        else {
            throw new Error('ID value unexpected');
        }
    };

    return <form id='crypto-input-form'>
        <div className='crypto-input-field'>
            <label htmlFor='dashboard_token' className='crypto-input-label-class'>
                Token
            </label>
            <input type='text' id='dashboard_token' onChange={e => changeEvent(e, ID['0'])}
                className='crypto-text-input-class' value={state.token} />
        </div>
        <div className='crypto-input-field'>
            <label htmlFor='dashboard_quantity' className='crypto-input-label-class'>
                Qty. Owned
            </label>
            <input type='number' id='dashboard_quantity' onChange={e => changeEvent(e, ID['1'])}
                className='crypto-text-input-class' value={state.quantity} />
        </div>
        <div id='crypto-button-container'>
            <button type='submit' id='dashboard_add_button'
                onClick={handleSubmit} disabled={loading}>{(edit || test) ? 'Save Asset' : 'Add Asset'}</button>
            <div id='add_asset_error' style={{ display: error ? 'block' : 'none' }}
            >Error: Please fill all the details</div>
        </div>
    </form>;
};
