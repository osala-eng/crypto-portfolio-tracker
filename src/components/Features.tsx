import React from 'react';
import './css/Features.css';

const Features = () => {
    return (
        <div id='features-container'>
            <h1 id='features_heading'>Features</h1>
            <div id='features_list'>
                <div className='feature-container'>
                    <p className='feature'>Add/Remove Crypto Assets</p>
                </div>
                <div className='feature-container'>
                    <p className='feature'>Track Prices</p>
                </div>
                <div className='feature-container'>
                    <p className='feature'>View PnL</p>
                </div>
            </div>
        </div>
    );
};

export default Features;
