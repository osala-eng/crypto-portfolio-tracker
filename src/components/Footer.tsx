import React from 'react';
import './css/Footer.css';
import linkedin from '../assets/linkedin.png';
import facebook from '../assets/facebook.png';
import twitter from '../assets/twitter.png';

const icon = 30;

const Footer = () => {
    return (
        <footer id='footer_container'>
            <div className='footer-div'></div>
            <div id='copyright_notice' className='footer-div'>
                Copyright Crypto Tracker
            </div>
            <div className='footer-div' id='sm_links'>
                <a href="//skillreactor.io" id='sm_linkedin' className='image-icos'>
                    <img src={linkedin} alt="linkedin icon" width={icon}
                        height={icon}  />
                </a>
                <a href="//skillreactor.io" id='sm_facebook' className='image-icos'>
                    <img src={facebook} alt="facebook icon" width={icon}
                        height={icon} />
                </a>
                <a href="//skillreactor.io" id='sm_linkedin' className='image-icos'>
                    <img src={twitter} alt='twitter icon' width={icon}
                        height={icon} />
                </a>
            </div>
        </footer>
    )
}

export default Footer;
