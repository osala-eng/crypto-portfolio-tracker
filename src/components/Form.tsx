import React from 'react';
import './Form.css';

export const Form = () => {
  return (
    <div className='form-container'>
        <form className='input-form'>
            <div>
                <label htmlFor='username_field' className='input-line'>
                    Username
                </label>
                <input type='text' name='username_field' id='username_field'
                className='input-line'/>
            </div>
            <input type='submit' value='Submit'  id='submit_button' className='input-line'/>
        </form>
    </div>
  )
};
