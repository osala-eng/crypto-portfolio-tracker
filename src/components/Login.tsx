import React, { useEffect, useState } from 'react';
// import { BackendUrl } from '../data/config';
import { ChangeEvent, ClickEvent, ID, MS } from '../data/types';
import './css/Register.css';
import { ErrorMsg, Loading } from './Messages';

const Login = () => {
    const [state, setState] = useState({
        username: '',
        password: ''
    });

    const [error, setError] = useState(false);
    const [loading, setLoading] = useState(false);

    /* istanbul ignore next */
    const handleState = (e: ChangeEvent, id: number) => {
        if (id === ID['1']) {
            setState({
                ...state, username: e.target.value
            });
        }
        else if (id === ID['2']) {
            setState({
                ...state, password: e.target.value
            });
        }
        else {
            throw new Error('Unexpected value');
        }
    };

    /* istanbul ignore next */
    const handleClick = async (e: ClickEvent) => {
        e.preventDefault();
        if (!state.password.length || !state.username.length) {
            setLoading(false);
            setError(true);
        }

        else {
            setLoading(true);
            await fetch('https://google.com', {
                // method: 'POST',
                // body: JSON.stringify(state)
            })
                .then(() => {
                    setLoading(false);
                    // if (res.status !== HTTP['201']) {
                    //     throw new Error('User registration failed');
                    // }
                })
                .catch(() => {
                    setLoading(false);
                    setError(true)
                });
        }
    };
    /* istanbul ignore next */
    useEffect(()=>{
        const time = setTimeout(()=>{
            setError(false);
        }, MS['10000']);

        return (()=>{
            clearTimeout(time);
        });

    }, [loading, error]);

    /* istanbul ignore next */
    const displayMsg = () => {
        if (loading){
            return <Loading />;
        }

        else if (error && !loading){
            return <ErrorMsg />;
        }

        else {
            return <></>;
        }
    };

    return (
        <div id='register-container'>
            <h1 id='login_heading'>Login</h1>
            <form id='login_form'>
                <div className='user-input' id='user-input-div'>
                    <label htmlFor='login_username_field' className='box-label'>username</label>
                    <input type='text' id='login_username_field' name='login_username_field'
                        className='box-input' onChange={(e) => handleState(e,ID['1'])}
                        data-testid={'login-user'} />
                </div>
                <div className='user-input' >
                    <label htmlFor='login_password_field' className='box-label'>password</label>
                    <input type='password' id='login_password_field' name='login_password_field'
                        className='box-input' onChange={(e) => handleState(e,ID['2'])}
                        data-testid={'login-pass'}/>
                </div>
                {displayMsg()}
                <div id='button_container'>
                    <button type='submit' id='login_button' onClick={handleClick}
                    data-testid={'login-btn'}>Login</button>
                </div>
            </form>
        </div>
    );
};

export default Login;
