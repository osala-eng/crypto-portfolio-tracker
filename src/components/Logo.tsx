import React from 'react';
import './css/Logo.css';
import { Link } from 'react-router-dom';

const logo = require('../assets/logo.png');



const logoSize = 150;
const Logo = ({logout=false, logoutFn}:
  {logout?: boolean, logoutFn?: ()=>void}) => {
  return (
    <div id='logo_disp' className='logo-container'>
      <Link to={'/'}>
        <img src={logo} alt='logo' id='logo' width={logoSize}
          height={logoSize} />
      </Link>
      <div id='name_box'>
        <h1 id='company_name'>Crypto Tracker</h1>
      </div>
      <div id='logout_link' style={{display: logout ? 'block' : 'none'}}
        onClick={logoutFn}>Logout</div>
    </div>
  );
};

export default Logo;
