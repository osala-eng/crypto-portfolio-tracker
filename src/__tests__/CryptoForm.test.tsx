import React from "react";
import {render, screen, waitFor} from '@testing-library/react';
import userEvent from "@testing-library/user-event";
import {CryptoForm} from '../components/CryptoForm';

jest.setTimeout(20000);
const options = {
    timeout: 15000
}
describe('Test crypto form', ()=> {
    render(<CryptoForm/>)
    test('Buttons and input operations', async() => {
        const token = await waitFor
            (() => screen.findByLabelText(/Token/i), options);
        const quantity = await waitFor
            (() => screen.findByLabelText(/Qty. Owned/i), options);
        const button = await waitFor
            (() => screen.findByText(/Add Asset/i), options);

        expect(token).toBeInTheDocument();
        expect(quantity).toBeInTheDocument();
        expect(button).toBeInTheDocument();

        userEvent.type(token, 'bitcoin');
        userEvent.type(quantity, '2');
        userEvent.click(button);

    })
})