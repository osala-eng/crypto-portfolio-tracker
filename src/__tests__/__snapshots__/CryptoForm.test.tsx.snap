// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create a snapshot with save button Snapshot test 1`] = `
<form
  id="crypto-input-form"
>
  <div
    className="crypto-input-field"
  >
    <label
      className="crypto-input-label-class"
      htmlFor="dashboard_token"
    >
      Token
    </label>
    <input
      className="crypto-text-input-class"
      id="dashboard_token"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="crypto-input-field"
  >
    <label
      className="crypto-input-label-class"
      htmlFor="dashboard_quantity"
    >
      Qty. Owned
    </label>
    <input
      className="crypto-text-input-class"
      id="dashboard_quantity"
      onChange={[Function]}
      type="number"
      value={NaN}
    />
  </div>
  <div
    id="crypto-button-container"
  >
    <button
      disabled={false}
      id="dashboard_save_button"
      onClick={[Function]}
      type="submit"
    >
      Save Asset
    </button>
    <div
      id="add_asset_error"
      style={
        Object {
          "display": "none",
        }
      }
    >
      Error: Please fill all the details
    </div>
  </div>
</form>
`;
